demo-app.py

```diff
@@ -241,38 +238,30 @@ def to_progressed_cursor(self) -> Cursor:
             # in progress.
             xid_next = min(xid_next, self.snapshot.xmax)
 
+        if self.previous_cursor.xid_at == xid_at:
+            # If xid_at did not move, then nothing from the xip_list will have been processed
+            xips_to_keep = self.previous_cursor.xip_list
+        else:
+            # If it did move, then at least some xip_list txids have been involved, and we don't need to
+            # retain anything less than xid_at, unless they're still in the snapshot, which
+            # get union-ed below. Now-committed txids from xip_list are processed in order, and the
+            # last one will become the new xid_at if we don't process all the changes.
+            xips_to_keep = [xip for xip in self.previous_cursor.xip_list if xip > xid_at]
```

Comment:
This expression seems semantically the same as the previous one, except for (a) some CPU and (b) I don't know if xip could == xid_at in the previous branch. And the 'some CPU' should be small since we're assuming in-memory datasets.

If (b) can't occur, then perhaps its worth highlighting the equivalence, or perhaps removing the special case.

Reply:
I'm not following your comment on how the expressions seem semantically the same.

`xid_at` is involved in the first union arm in the query with a `last_modified_txid = $xid_at` filter.

`xip_list` is involved in the second union arm, like `WHERE ... last_modified_txid = ANY($xip_list) ORDER BY last_modified_txid ASC, tiebreaker`.

If the second union arm fills up the batch, the txid of the last row becomes the next `xid_at`. Any txids in `xip_list` following it must be retained. Any txid smaller than it can be assumed to have been done already, due to the filter and the ORDER BY.